Add bit shift intrinsic functions

jrdodds opened the pull request:
Fixes #8551

### Context
The intrinsic functions include functions for bit And, Or, Xor, and Not operators but not for the bit shift operators.

### Changes Made
Added functions LeftShift, RightShift, and RightShiftUnsigned. The functions expect values of type `int` (32 bit) which matches the existing bit operator functions. `long` (64 bit) is not supported by any of the bit operator intrinsic functions.

### Testing
Tested on Windows 11 and macOS 12 Monterey.

Added to Expander_Tests.cs and ran unit tests.

Created and ran BitShift.proj project file with the following content:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">

  <Target Name="Test">
    <PropertyGroup>
      <LeftShiftTest>$([MSBuild]::LeftShift(1, 2))</LeftShiftTest>
      <RightShiftTest>$([MSBuild]::RightShift(-8, 2))</RightShiftTest>
      <RightShiftUnsignedTest>$([MSBuild]::RightShiftUnsigned(-8, 2))</RightShiftUnsignedTest>
    </PropertyGroup>

    <Message Text="LeftShiftTest: $(LeftShiftTest)" />
    <Message Text="RightShiftTest: $(RightShiftTest)" />
    <Message Text="RightShiftUnsignedTest: $(RightShiftUnsignedTest)" />
  </Target>

</Project>

```

The project produced the following output (which matched the expected results):

```
Test:
  LeftShiftTest: 4
  RightShiftTest: -2
  RightShiftUnsignedTest: 1073741822
```

### Notes

Forgind commented on src/Build/Evaluation/IntrinsicFunctions.cs:
```diff
@@ -165,6 +165,21 @@ internal static int BitwiseNot(int first)
+
+        internal static int RightShiftUnsigned(int operand, int count)
+        {
+            return operand >>> count;
```
I wasn't aware this existed. Always nice to learn something new!

Forgind reviewed (approved):
Thanks!

rainersigwald commented:
In the past we've added any method that is known to be called in the wild to the fast path, but haven't required it, and I suspect the bitwise operations were missed, rather than intentionally omitted.

jrdodds commented:
I added a "fast path" for each of the bit operator functions:

- BitwiseOr
- BitwiseAnd
- BitwiseXor
- BitwiseNot
- LeftShift
- RightShift
- RightShiftUnsigned

In `Expander_Tests` I separated the bit operator functions into their own test method from the math functions, and I changed from Xunit Assert's to Shouldly. (I was motivated to make the unit test change because I had a typo and the Xunit Assert message didn't include the name of the failing function.)

ladipro commented:
> @rainersigwald, the existing bitwise operations are not on the fast path either. Presumably because they're not expected to be used very often and, while making the first call faster, every such special case `if` incurs a small steady state perf hit.

Make sense. Also, source generators for the win! It should be a fun project to automatically create efficient lookup tables at compile time.

rainersigwald commented:
> Also, source generators for the win! It should be a fun project to automatically create efficient lookup tables at compile time.

I have been fantasizing about doing this for several years now!

jrdodds commented:
I'm looking at the 'Fast Path' `TryExecuteWellKnownFunction` code for other functions and I have some questions.

`IntrinsicFunctions.cs` includes `Add` functions for `double` and `int`:
https://github.com/dotnet/msbuild/blob/b84faa7d0dfb42daedca5c244f73f0b07e1135d7/src/Build/Evaluation/IntrinsicFunctions.cs#L40-L55

The `TryExecuteWellKnownFunction` method in `Expander.cs` tests for but doesn't use the `Add` functions:
https://github.com/dotnet/msbuild/blob/b84faa7d0dfb42daedca5c244f73f0b07e1135d7/src/Build/Evaluation/Expander.cs#L3893-L3900

`Subtract`, `Multiply`, and `Divide` have the same pattern. The operator is used instead of invoking the function.

There are `Modulo` functions for `double` and `int` and they are not in the `TryExecuteWellKnownFunction` method.

Shouldn't the `TryExecuteWellKnownFunction` method always be invoking the appropriate member of `IntrinsicFunctions` (and never directly using the operator)?

It seems that currently `Add`, `Subtract`, `Multiply`, and `Divide` are always performed with `double`s. How are the `int` and `double` overloads intended to resolve? Does the 'slow path' resolve the type overload?

Forgind commented:
> Shouldn't the TryExecuteWellKnownFunction method always be invoking the appropriate member of IntrinsicFunctions (and never directly using the operator)?

I don't think this really matters. From a style perspective, it is better to have them all look the same, that is, all call the appropriate member of IntrinsicFunctions as you suggest. From the user's perspective, they're equivalent, and if the call isn't inlined, calling '+' directly may even be (negligibly) faster.

> It seems that currently Add, Subtract, Multiply, and Divide are always performed with doubles. How are the int and double overloads intended to resolve? Does the 'slow path' resolve the type overload?

There aren't really "ints," "doubles," etc. in MSBuild; everything is just strings. Even things like [TryConvertToInt](https://github.com/dotnet/msbuild/blob/b84faa7d0dfb42daedca5c244f73f0b07e1135d7/src/Build/Evaluation/Expander.cs#L4413) are kinda misleading, as far as I know, because 'value' won't ever be a double or an int; it'll always be a string, and we'll always parse it as such. With that in mind, if you pass Add two ints, they just turn into doubles, get added, and get converted back to a string that looks like an integer, so people generally don't care too much about whether the arithmetic was actually executed with ints or doubles. Since all ints are doubles, though, we do get them in the fast path.

jrdodds commented:
@Forgind It sounds like the int overloads of Add, Subtract, Multiply, Divide, and Modulo should be removed because they are never used.

Forgind commented:
I wouldn't be opposed to that. The only ways I can think of that someone could actually call them is if they use reflection or MSBuildLocator to specifically load our IntrinsicFunctions class then call the int overload...neither feels at all worthwhile to me, since C# programmers don't need a special function to call something built-in. Can you think of any legitimate reason someone might want to use those overload @rainersigwald?

rainersigwald commented:
@jrdodds I'm not sure of the impact of some of those removals, and it's late enough in the 17.6 release cycle that I'm nervous about removing them. Can you separate them into a different PR that we can think about for a bit longer? Adding the shift operators and the fast paths is in good shape, though.

jrdodds commented:
@rainersigwald This PR has been updated.
